### PR TITLE
Task-54926 : Anti-Brute force

### PR DESCRIPTION
**Pull request by rdenarie**
This commit add the feature to prevent brute force attack on login

After this fix, when a user fail to log X times, his account is locked during Y minutes. He also receive a notification mail explaining the problem, and he is invited to make a forgot password request

### CODE REVIEW

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/Config.java [hunk 1]
```diff
@@ -92,6 +92,9 @@
 
     private boolean updateLastLoginTimeAfterAuthentication = true;
 
+    private int maxBadPasswordCount;
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/Config.java]: not  bad, wrong

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 2]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+        }
+        int currentNbFail = profile.getAttribute(BAD_PASSWORD_COUNT) != null ?
+                            Integer.parseInt(profile.getAttribute(BAD_PASSWORD_COUNT)) : 0;
+        currentNbFail++;
+        profile.setAttribute(BAD_PASSWORD_COUNT, String.valueOf(currentNbFail));
+        Instant now = Instant.now();
+        profile.setAttribute(BAD_PASSWORD_TIME, String.valueOf(now.toEpochMilli()));
+        orgService.getUserProfileHandler().saveUserProfile(profile, true);
+
+        if (currentNbFail >= orgService.getConfiguration().getMaxBadPasswordCount()) {
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]: how about sending an analytics event also?

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 3]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+                   orgService.getConfiguration().getMaxBadPasswordCount(),
+                   now,
+                   orgService.getConfiguration().getBlockingTime(),
+                   now.plus(orgService.getConfiguration().getBlockingTime(), ChronoUnit.MINUTES));
+          ExoContainer container = ExoContainerContext.getCurrentContainer();
+          PasswordRecoveryService passwordRecoveryService = container.getComponentInstanceOfType(PasswordRecoveryService.class);
+          passwordRecoveryService.sendAccountLockedEmail(user, Locale.ENGLISH);
+
+        } else {
+          log.warn("service=login" + " operation=login" + " status=ko"
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]: how about adding an analytics event also?

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/Config.java (hunk 1 above)
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/Config.java]:
```suggestion
    private int maxAuthenticationAttempts;
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/Config.java [hunk 4]
```diff
@@ -376,4 +379,20 @@ public boolean isUseCache() {
     public void setUseCache(boolean useCache) {
       this.useCache = useCache;
     }
+
+    public int getMaxBadPasswordCount() {
+        return maxBadPasswordCount;
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/Config.java]:
```suggestion
        return maxAuthenticationAttempts;
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/Config.java [hunk 5]
```diff
@@ -376,4 +379,20 @@ public boolean isUseCache() {
     public void setUseCache(boolean useCache) {
       this.useCache = useCache;
     }
+
+    public int getMaxBadPasswordCount() {
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/Config.java]:
```suggestion
    public int getMaxAuthenticationAttempts() {
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/Config.java [hunk 6]
```diff
@@ -376,4 +379,20 @@ public boolean isUseCache() {
...
+        return maxBadPasswordCount;
+    }
+
+    public void setMaxBadPasswordCount(int maxBadPasswordCount) {
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/Config.java]:
```suggestion
    public void setMaxAuthenticationAttempts(int maxAuthenticationAttempts) {
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/Config.java [hunk 7]
```diff
@@ -376,4 +379,20 @@ public boolean isUseCache() {
...
+        this.maxBadPasswordCount = maxBadPasswordCount;
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/Config.java]:
```suggestion
        this.maxAuthenticationAttempts = maxAuthenticationAttempts;
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 8]
```diff
@@ -68,6 +71,10 @@
 
   public static final DateFormat  dateFormat           = DateFormat.getInstance();
 
+  private static final String BAD_PASSWORD_COUNT = "badPasswordCount";
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
  private static final String AUTHENTICATION_ATTEMPTS = "authenticationAttempts";
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 9]
```diff
@@ -68,6 +71,10 @@
...
+
+  private static final String BAD_PASSWORD_TIME = "badPasswordTime";
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
  private static final String LATEST_AUTH_TIME = "latestAuthFailureTime";
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 10]
```diff
@@ -370,15 +377,24 @@ public boolean authenticate(String username, String password) throws Exception {
...
 
-      if (user.isInternalStore()) {
-        authenticated = authenticateDB(user, password);
-        if (authenticated) {
-          return true;
+      if (!isAccountLocked(user)) {
+        if (user.isInternalStore()) {
+          authenticated = authenticateDB(user, password);
+          if (authenticated) {
+            resetBadPassordCount(user.getUserName());
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
            resetAuthenticationAttempts(user.getUserName());
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 11]
```diff
@@ -370,15 +377,24 @@ public boolean authenticate(String username, String password) throws Exception {
...
         }
       }
     }
 
-    return authenticateExternal(username, password);
+    authenticated = authenticateExternal(username, password);
+    if (!authenticated) {
+      saveLastLoginFail(username);
+    } else {
+      resetBadPassordCount(username);
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
      resetAuthenticationAttempts(username);
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 12]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+
+  private void saveLastLoginFail(String username) {
+    try {
+      User user = orgService.getUserHandler().findUserByName(username);
+      if (user != null) {
+        UserProfile profile = orgService.getUserProfileHandler().findUserProfileByName(username);
+        if (profile == null) {
+          profile = orgService.getUserProfileHandler().createUserProfileInstance(username);
+        }
+        int currentNbFail = profile.getAttribute(BAD_PASSWORD_COUNT) != null ?
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
        int currentNbFail = profile.getAttribute(AUTHENTICATION_ATTEMPTS) != null ?
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 13]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+                            Integer.parseInt(profile.getAttribute(BAD_PASSWORD_COUNT)) : 0;
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
                            Integer.parseInt(profile.getAttribute(AUTHENTICATION_ATTEMPTS)) : 0;
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 14]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+        currentNbFail++;
+        profile.setAttribute(BAD_PASSWORD_COUNT, String.valueOf(currentNbFail));
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
        profile.setAttribute(AUTHENTICATION_ATTEMPTS, String.valueOf(currentNbFail));
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 15]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+        Instant now = Instant.now();
+        profile.setAttribute(BAD_PASSWORD_TIME, String.valueOf(now.toEpochMilli()));
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
        profile.setAttribute(LATEST_AUTH_TIME, String.valueOf(now.toEpochMilli()));
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 16]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+        orgService.getUserProfileHandler().saveUserProfile(profile, true);
+
+        if (currentNbFail >= orgService.getConfiguration().getMaxBadPasswordCount()) {
+          log.warn("service=login" + " operation=login" + " status=ko"
+              + " parameters=\"username:{}, badPasswordCount:{}, maxBadPasswordCount:{}, badPasswordTime={}, "
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
              + " parameters=\"username:{}, authenticationAttempts:{}, maxAuthenticationAttempts:{}, latestAuthenticationTime={}, "
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 17]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+              + "lockTimeInMinutes={}, unlockTime={}\"" + " error_msg=\"Account is locked\"",
+                   user.getUserName(),
+                   currentNbFail,
+                   orgService.getConfiguration().getMaxBadPasswordCount(),
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
                   orgService.getConfiguration().getMaxAutthenticationAttempts(),
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 18]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+                   now,
+                   orgService.getConfiguration().getBlockingTime(),
+                   now.plus(orgService.getConfiguration().getBlockingTime(), ChronoUnit.MINUTES));
+          ExoContainer container = ExoContainerContext.getCurrentContainer();
+          PasswordRecoveryService passwordRecoveryService = container.getComponentInstanceOfType(PasswordRecoveryService.class);
+          passwordRecoveryService.sendAccountLockedEmail(user, Locale.ENGLISH);
+
+        } else {
+          log.warn("service=login" + " operation=login" + " status=ko"
+              + " parameters=\"username:{}, badPasswordCount:{}, badPasswordTime:{}, maxBadPasswordCount:{}\""
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
              + " parameters=\"username:{}, authenticationAttempts:{}, latestAuthenticationTime:{}, maxAuthenticationAttempts:{}\""
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 19]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+              + " error_msg=\"Login failed\"",
+                   username,
+                   currentNbFail,
+                   now,
+                   orgService.getConfiguration().getMaxBadPasswordCount());
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
                   orgService.getConfiguration().getMaxAuthenticationAttempts());
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 20]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+        }
+      }
+    } catch (Exception e) {
+      log.error("Unable to get gatein user profile for user {}", username, e);
+    }
+
+  }
+
+  private void resetBadPassordCount(String username) {
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
  private void resetAuthenticationAttempts(String username) {
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 21]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+    try {
+      User user = orgService.getUserHandler().findUserByName(username);
+      if (user != null) {
+        UserProfile profile = orgService.getUserProfileHandler().findUserProfileByName(username);
+        if (profile == null) {
+          profile = orgService.getUserProfileHandler().createUserProfileInstance(username);
+        }
+        profile.setAttribute(BAD_PASSWORD_COUNT, String.valueOf(0));
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
        profile.setAttribute(AUTHENTICATION_ATTEMPTS, String.valueOf(0));
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 22]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+        orgService.getUserProfileHandler().saveUserProfile(profile, true);
+        if (log.isDebugEnabled()) {
+          log.debug("service=login"
+                       + " operation=login"
+                       + " status=ok"
+                       + " parameters=\"username:{}, badPasswordCount:{}, maxBadPasswordCount:{}\"",
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
                       + " parameters=\"username:{}, authenticationAttempts:{}, maxAuthenticationAttempts:{}\"",
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 23]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+  }
+
+  private boolean isAccountLocked(User user) throws AccountTemporaryLockedException{
+    try {
+      UserProfile profile = orgService.getUserProfileHandler().findUserProfileByName(user.getUserName());
+      if (profile == null) {
+        //if there is no userProfile, the user have never fail his login, we do no lock him
+        return false;
+      }
+      int currentNbFail = profile.getAttribute(BAD_PASSWORD_COUNT) != null ?
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
      int currentNbFail = profile.getAttribute(AUTHENTICATION_ATTEMPTS) != null ?
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 24]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+                          Integer.parseInt(profile.getAttribute(BAD_PASSWORD_COUNT)) : 0;
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
                          Integer.parseInt(profile.getAttribute(AUTHENTICATION_ATTEMPTS)) : 0;
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 25]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+      Instant badPasswordTime = Instant.ofEpochMilli(Long.parseLong(profile.getAttribute(BAD_PASSWORD_TIME)));
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
      Instant badPasswordTime = Instant.ofEpochMilli(Long.parseLong(profile.getAttribute(AUTHENTICATION_ATTEMPTS)));
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 26]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+      if (currentNbFail >= orgService.getConfiguration().getMaxBadPasswordCount()
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
      if (currentNbFail >= orgService.getConfiguration().getMaxAuthenticationAttempts()
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 27]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+          &&
+          badPasswordTime.plus(orgService.getConfiguration().getBlockingTime(), ChronoUnit.MINUTES).isAfter(Instant.now())) {
+
+        log.warn("service=login"
+                     + " operation=login"
+                     + " status=ko"
+                     + " parameters=\"username:{}, badPasswordCount:{}, maxBadPasswordCount:{}, badPasswordTime={}, "
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
                     + " parameters=\"username:{}, authenticationAttempts:{}, maxAuthenticationAttempts:{}, latestAuthenticationTime={}, "
```

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java [hunk 28]
```diff
@@ -915,4 +931,113 @@ private boolean filterDisabledUsersInQueries() {
...
+                     + "lockTimeInMinutes={}, unlockTime={}\""
+            + " error_msg=\"Account is locked\"",
+                 user.getUserName(),
+                 currentNbFail,
+                 orgService.getConfiguration().getMaxBadPasswordCount(),
```
Comment (plamarque) [inline component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]:
```suggestion
                 orgService.getConfiguration().getMaxAuthenticationAttempts(),
```

File: web/portal/src/main/webapp/WEB-INF/conf/organization/idm-configuration.xml [hunk 29]
```diff
@@ -232,25 +232,35 @@
...
-              <boolean>${exo.idm.user.updateLastLoginTime:true}</boolean>
-            </field>
+          <!-- If true, the property lastLoginTime of user will be updated after successful authentication of this user to portal.
+               If false, the property won't be updated, which could have performance improvement in systems with many concurrent user logins -->
+          <field name="updateLastLoginTimeAfterAuthentication">
+            <boolean>${exo.idm.user.updateLastLoginTime:true}</boolean>
+          </field>
+
+          <!-- This property define how many fail attempts a user can do before his account is blocked -->
+          <field name="maxBadPasswordCount">
```
Comment (plamarque) [inline web/portal/src/main/webapp/WEB-INF/conf/organization/idm-configuration.xml]:
```suggestion
          <field name="maxAuthenticaitonAttempts">
```

File: web/portal/src/main/webapp/WEB-INF/conf/organization/idm-configuration.xml [hunk 30]
```diff
@@ -232,25 +232,35 @@
...
+            <int>${exo.idm.antibruteforce.maxBadPassword:5}</int>
```
Comment (plamarque) [inline web/portal/src/main/webapp/WEB-INF/conf/organization/idm-configuration.xml]:
```suggestion
            <int>${exo.idm.antibruteforce.maxAuthenticationAttempts:5}</int>
```

**Review by plamarque — CHANGES_REQUESTED** [review]
I suggested new language.  authenticationAttempts instead of bad password count (which seems more gramatically correct to me). And latestAuthenticationTime instead badPasswordTime

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java (hunk 2 above)
Comment (rdenarie) [reply component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]: analytics added for failed login, and locked account

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java (hunk 8 above)
Comment (rdenarie) [reply component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]: change to wrongPasswordCount

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java (hunk 10 above)
Comment (rdenarie) [reply component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]: done

File: component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java (hunk 12 above)
Comment (rdenarie) [reply component/identity/src/main/java/org/exoplatform/services/organization/idm/UserDAOImpl.java]: done

**Review by plamarque — DISMISSED** [review]
LGTM

**Review by montassarhamadi — APPROVED** [review]

**Pull request merged**